lib/spack/spack/cmd/compiler.py

```diff
@@ -73,7 +73,8 @@ def compiler_find(args):
        add them to Spack's configuration.
 
     """
-    paths = args.add_paths
+    # None signals spack.compiler.find_compilers to use its default logic
+    paths = args.add_paths or None
```

Comment:
@adrienbernede @scheibelp These two lines are the fix, all the rest is code needed to set up a regression test. As an aside, caches at module level make it much more difficult to diagnose test failures - maybe we could have a look how to rework them sooner or later :sweat_smile:

Reply:
Normally I'd have preferred to make the function more robust (e.g. treat `None` and `[]` the same). Is there a reason you are handling this in the caller?

Reply:
There was an explicit request to change the semantic of `find_compilers` in https://github.com/spack/spack/pull/10190#discussion_r291039122, but I forgot to update the line above (and there was no test that failed because of this).

Reply:
I think not passing anything (None) and passing an empty list are two different things — one says do the default thing, the other could’ve been passed programmatically, and I would expect if the caller passed [], it would be surprising to see the routine use PATH.  See the comment on #10190.

Reply:
Whoops I did overlook that this was explicitly requested so no need to change that.

IMO it would be worth clarifying the behavior for an empty list in the docstring for `find_compilers`, since APIs based on treating `None` and objects where `bool(x) == False` can be troublesome (although that looks like the way to go for this function).

Reply:
There's a mention of that behavior in the `Args` section:

https://github.com/spack/spack/blob/c241e2ed81c97067f710dc216ff31f397efef259/lib/spack/spack/compilers/__init__.py#L194-L196

Is it worth adding something also to brief description above?